File: tests/ml_cpp_test.cpp
/* 
	g++ -I /usr/local/include -L /usr/local/lib ml_cpp_test.cpp embedded.pb.cc primitives.pb.cc -lprotobuf -pthread
*/

#include <iostream>
#include <fstream>
#include <chrono>

#include "embedded.pb.h"
#include "primitives.pb.h"
#include <google/protobuf/util/message_differencer.h>

using namespace std;

// Parsed messages, as decoded from the *_ml_encoded.dat files.
SmallPrimitiveCollection parsed_small_collection;
FullPrimitiveCollection parsed_full_collection;
M1 parsed_embedded_message;

// Original messages (values set by hand, they match ML test values.)
SmallPrimitiveCollection original_small_collection;
FullPrimitiveCollection original_full_collection;
M1 original_embedded_message;

// Build functions build the original objects (values added by hand).
void buildSimpleCollection() {
	original_small_collection.set_my_int32(-1);
	original_small_collection.set_my_int64(221921921929121);
	original_small_collection.set_my_string("hello");
}

void buildFullCollection() {
	original_full_collection.set_my_int32(-1);
	original_full_collection.set_my_uint32(3232321);
	original_full_collection.set_my_int32(-1);
	original_full_collection.set_my_int64(221921921929121);
	original_full_collection.set_my_uint32(3232321);
	original_full_collection.set_my_uint64(9977533344322);
	original_full_collection.set_my_sint32(-123);
	original_full_collection.set_my_sint64(-221921921929121);
	original_full_collection.set_my_fixed32(-323231);
	original_full_collection.set_my_fixed64(221921921929121);
	original_full_collection.set_my_sfixed32(-67565);
	original_full_collection.set_my_sfixed64(69595969659659);
	original_full_collection.set_my_bool(true);
	char arr[] = {0x1,0xFF,0x23}; 
	string s(arr);
	original_full_collection.set_my_bytes(s);
	original_full_collection.set_my_string("hello");
	original_full_collection.set_my_float((float)219129.2111);
	original_full_collection.set_my_double((double)-8382832823.392932211);
}

void buildEmbeddedMessage() {
	M1_M2* m2_1 = original_embedded_message.mutable_message2_1();
	M1_M2* m2_2 = original_embedded_message.mutable_message2_2();
	m2_1->set_name("random");
	m2_1->set_id(121);
	m2_2->set_name("my name");
	m2_2->set_id(223);
	
	M3* m3_1 = original_embedded_message.add_message3_list();
	m3_1->set_name("some name");
	M3* m3_2 = original_embedded_message.add_message3_list();
	m3_2->set_name("another");
	M3* m3_3 = original_embedded_message.add_message3_list();
	m3_3->set_name("yet another");
}

// Functions to parse messages encoded by ML library.
void readSimpleCollectionFromFile() {
	fstream input("smallcollection_ml_encoded.dat", ios::in | ios::binary);
	if (!parsed_small_collection.ParseFromIstream(&input)) {
		cerr << "Failed to parse SmallPrimitiveCollection instance from smallcollection_ml_encoded.dat\n";
	}
}

void readFullCollectionFromFile() {
	fstream input("fullcollection_ml_encoded.dat", ios::in | ios::binary);
	if (!parsed_full_collection.ParseFromIstream(&input)) {
		cerr << "Failed to parse FullPrimitiveCollection instance from fullcollection_ml_encoded.dat\n";
	}
}

void readEmbeddedMessageFromFile() {
	fstream input("embedded_ml_encoded.dat", ios::in | ios::binary);
	if (!parsed_embedded_message.ParseFromIstream(&input)) {
		cerr << "Failed to parse embedded message instance from embedded_ml_encoded.dat\n";
	}
}

// Compares each field in between the parsed and original messages.
void compareSimpleCollection() {
	string error_string;
	google::protobuf::util::MessageDifferencer differencer;

	differencer.ReportDifferencesToString(&error_string);

	if (differencer.Compare(parsed_small_collection, 
		original_small_collection)) {
		cout << "SUCCESS: Small collection messages match!\n";
	} else {
		cout << "ERROR: Small collection messages don't match!\n";
		cerr << error_string << "\n";
	}
}

void compareFullCollection() {
	string error_string;
	google::protobuf::util::MessageDifferencer differencer;

	differencer.ReportDifferencesToString(&error_string);

	if (differencer.Compare(parsed_full_collection, original_full_collection)) {
		cout << "SUCCESS: Full collection messages match!\n";
	} else {
		cout << "ERROR: Full collection messages don't match!\n";
		cerr << error_string << "\n";
	}
}

// Compares lists as sets and outputs any differences.
void compareEmbeddedMessage() {
	string error_string;
	google::protobuf::util::MessageDifferencer differencer;

	differencer.set_repeated_field_comparison(
		google::protobuf::util::MessageDifferencer::AS_SET);
	differencer.ReportDifferencesToString(&error_string);

	if (differencer.Compare(parsed_embedded_message, 
		original_embedded_message)) {
		cout << "SUCCESS: Embedded messages match!\n";
	} else {
		cout << "ERROR: Embedded messages don't match!\n";
		cerr << error_string << "\n";
	}
}

void serialize_small_collection_test() {
	fstream output("smallcollection_cpp_encoded.dat", ios::out | ios::trunc 
		| ios::binary);
	original_small_collection.SerializeToOstream(&output);
}

void serialize_full_collection_test() {
	fstream output("fullcollection_cpp_encoded.dat", ios::out | ios::trunc | ios::binary);
	original_full_collection.SerializeToOstream(&output);
}

void serialize_embedded_message_test() {
	fstream output("embedded_cpp_encoded.dat", ios::out | ios::trunc | ios::binary);
	original_embedded_message.SerializeToOstream(&output);
}

int main() {
	buildSimpleCollection();
	buildFullCollection();
	buildEmbeddedMessage();

	readSimpleCollectionFromFile();
	readFullCollectionFromFile();
	readEmbeddedMessageFromFile();

	compareSimpleCollection();
	compareFullCollection();
	compareEmbeddedMessage();

	serialize_small_collection_test();
	serialize_full_collection_test();
	serialize_embedded_message_test();
	return 0;
}

File: ml_cpp_test.ml
(* Test for decoding (attempts to decode a data file encoded using the C++
library. *)

use "tests/primitives.pb.ml";
use "tests/embedded.pb.ml";
use "tests/test_helpers.ml";

exception invalidTest

(* Writes to file_name bytes from the Word8Vector argument. Useful for porting 
benchmark test data to other languages for testing *)
fun writeToFile file_name raw_bytes = 
let
	val ostream = BinIO.openOut file_name
in
	(BinIO.output (ostream, raw_bytes);
	BinIO.flushOut ostream)
end

(* Returns a Word8Vector from the raw bits in the file *)
fun getByteInputStream file_name = 
let
	val raw_vector = BinIO.inputAll (BinIO.openIn file_name)
	val input_stream = ByteInputStream.fromVector raw_vector
in
	input_stream
end

fun smallCollection () = 
let
	val builder = SmallPrimitiveCollection.Builder.init();
	val record = (
		SmallPrimitiveCollection.Builder.set_my_int32 (builder, ~1);
		SmallPrimitiveCollection.Builder.set_my_int64 (builder, 221921921929121);
		SmallPrimitiveCollection.Builder.set_my_string (builder, "hello");
		SmallPrimitiveCollection.Builder.build builder
	) 
in
	record
end

fun fullCollection () = 
let 
	val builder = FullPrimitiveCollection.Builder.init();
	val record = (
		FullPrimitiveCollection.Builder.set_my_int32 (builder, ~1);
		FullPrimitiveCollection.Builder.set_my_int64 (builder, 221921921929121);
		FullPrimitiveCollection.Builder.set_my_uint32 (builder, 3232321);
		FullPrimitiveCollection.Builder.set_my_uint64 (builder, 9977533344322);
		FullPrimitiveCollection.Builder.set_my_sint32 (builder, ~123);
		FullPrimitiveCollection.Builder.set_my_sint64 (builder, ~221921921929121);
		FullPrimitiveCollection.Builder.set_my_fixed32 (builder, ~323231);
		FullPrimitiveCollection.Builder.set_my_fixed64 (builder, 221921921929121);
		FullPrimitiveCollection.Builder.set_my_sfixed32 (builder, ~67565);
		FullPrimitiveCollection.Builder.set_my_sfixed64 (builder, 69595969659659);
		FullPrimitiveCollection.Builder.set_my_bool (builder, true);
		FullPrimitiveCollection.Builder.set_my_bytes (builder, Word8Vector.fromList [0wx1,0wxFF,0wx23]);
		FullPrimitiveCollection.Builder.set_my_string (builder, "hello");
		(* 219129.21875 is the cast to double of the next representable float
		of 219129.2111 *)
		FullPrimitiveCollection.Builder.set_my_float (builder, 219129.2111);
		FullPrimitiveCollection.Builder.set_my_double (builder, ~8382832823.392932211);
		FullPrimitiveCollection.Builder.build builder
	) 
in
	record
end

fun embeddedMessage () = 
let 
	(* Build 3 M3 messages to add to repeated field *)
	val m3_builder1 = M3.Builder.set_name (M3.Builder.init(), "some name");
	val m3_builder2 = M3.Builder.set_name (M3.Builder.init(), "another");
	val m3_builder3 = M3.Builder.set_name (M3.Builder.init(), "yet another");
	val m3_1 = M3.Builder.build m3_builder1
	val m3_2 = M3.Builder.build m3_builder2
	val m3_3 = M3.Builder.build m3_builder3

	val m2_builder1 = M1.M2.Builder.set_name (M1.M2.Builder.init(), "random");
	val m2_builder1 = M1.M2.Builder.set_id (m2_builder1, 121);
	val m2_builder2 = M1.M2.Builder.set_name (M1.M2.Builder.init(), "my name");
	val m2_builder2 = M1.M2.Builder.set_id (m2_builder2, 223);
	val m2_1 = M1.M2.Builder.build m2_builder1;
	val m2_2 = M1.M2.Builder.build m2_builder2;

	val m1_builder = M1.Builder.init();
	val m1_record = (
		M1.Builder.set_message2_1 (m1_builder, m2_1);
		M1.Builder.set_message2_2 (m1_builder, m2_2);
		M1.Builder.add_message3_list (m1_builder, m3_1);
		M1.Builder.add_message3_list (m1_builder, m3_2);
		M1.Builder.add_message3_list (m1_builder, m3_3);
		M1.Builder.build m1_builder
	)
in 
	m1_record
end

fun decode 0 = 
	let
		val input_stream = getByteInputStream "tests/smallcollection_cpp_encoded.dat";
		val (dec, parse_result) = SmallPrimitiveCollection.decodeToplevel input_stream
	in
		dec = (smallCollection())
	end
| decode 1 = 
	let
		val input_stream = getByteInputStream "tests/fullcollection_cpp_encoded.dat";
		val (dec, parse_result) = FullPrimitiveCollection.decodeToplevel input_stream
		val record = fullCollection()
	in
		(#my_int32 record) = (#my_int32 dec) andalso
		(#my_int64 record) = (#my_int64 dec) andalso
		(#my_uint32 record) = (#my_uint32 dec) andalso
		(#my_uint64 record) = (#my_uint64 dec) andalso
		(#my_sint32 record) = (#my_sint32 dec) andalso
		(#my_sint64 record) = (#my_sint64 dec) andalso
		(#my_fixed32 record) = (#my_fixed32 dec) andalso
		(#my_fixed64 record) = (#my_fixed64 dec) andalso
		(#my_sfixed32 record) = (#my_sfixed32 dec) andalso
		(#my_sfixed64 record) = (#my_sfixed64 dec) andalso
		(#my_bool record) = (#my_bool dec) andalso
		(#my_bytes record) = (#my_bytes dec) andalso
		(#my_string record) = (#my_string dec) andalso
		(* Manually comparing the float value as precision is lost, but 
		value is deterministic: the cast to double of the next representable
		float of the original value.*)
		(Real.== (219129.21875, Option.valOf(#my_float dec))) andalso
		(Real.== (Option.valOf (#my_double record), Option.valOf(#my_double dec)))
	end
| decode 2 = 
	let
		val input_stream = getByteInputStream "tests/embedded_cpp_encoded.dat";
		val (dec, parse_result) = M1.decodeToplevel input_stream
		val m1_record = embeddedMessage()
	in
		list_eq((#message3_list) m1_record, (#message3_list) dec) andalso
		(#message2_1) m1_record = (#message2_1) dec andalso
		(#message2_2) m1_record = (#message2_2) dec
	end
| decode n = raise invalidTest

fun encode 0 = 
	(* Simple test for message of type SmallPrimitiveCollection *)
	let
		val record = smallCollection ();
		val enc = SmallPrimitiveCollection.encodeToplevel record
	in
		(print("Writing SmallPrimitiveCollection to \"tests/smallcollection_ml_encoded.dat\".\n");
		 writeToFile "tests/smallcollection_ml_encoded.dat" enc)
	end
  | encode 1 = (* Test for message of type FullPrimitiveCollection, with floats. *)
	let
		val builder = FullPrimitiveCollection.Builder.init();
		val record = fullCollection ();
		val enc = FullPrimitiveCollection.encodeToplevel record
	in
		(print("Writing FullPrimitiveCollection to \"tests/fullcollection_ml_encoded.dat\".\n");
		 writeToFile "tests/fullcollection_ml_encoded.dat" enc)
	end
  | encode 2 = 	(* Simple test for message of type M1 in embedded.proto *)
	let
		val m1_record = embeddedMessage();
		val enc = M1.encodeToplevel m1_record
	in
		
		(print("Writing M1 embedded message to \"tests/embedded_ml_encoded.dat\".\n");
		 writeToFile "tests/embedded_ml_encoded.dat" enc)
	end
  | encode n = raise invalidTest

val lastTest = 2;

fun runEncoding () = 
let
	fun run i = if i <= lastTest then (encode i; run (i+1)) else ()
in
	run 0
end

fun runDecoding () = 
let
	fun run i = 
		if i > lastTest then (true, "TESTS PASS! All messages were decoded correctly!")
		else if (decode i) then (run (i+1))
		else (false, "Test " ^ Int.toString(i) ^ " failed")
	val (b, message) = run 0
in
	print(message^"\n")
end

fun main() = runDecoding ();